docs/dotnet/dotnet-http-metrics.md

```diff
@@ -0,0 +1,257 @@
+# Semantic Conventions for HTTP-relevant metrics emitted by .NET and ASP.NET Core
+
+**Status**: [Experimental][DocumentStatus]
+
+This document defines semantic conventions for HTTP metrics emitted by .NET components and runtime.
+
+**Disclaimer:** These are initial .NET metric instruments available in .NET 8 but more may be added in the future.
+
+<!-- toc -->
+
+- [HTTP client](#http-client)
+  * [Metric: `http.client.connections.usage`](#metric-httpclientconnectionsusage)
+  * [Metric: `http.client.connection.duration`](#metric-httpclientconnectionduration)
+  * [Metric: `http.client.duration`](#metric-httpclientduration)
+  * [Metric: `http.client.active_requests`](#metric-httpclientactive_requests)
+  * [Metric: `http.client.failed_requests`](#metric-httpclientfailed_requests)
+- [HTTP server](#http-server)
+  * [Metric: `http.server.duration`](#metric-httpserverduration)
+  * [Metric: `http.server.active_requests`](#metric-httpserveractive_requests)
+  * [Metric: `http.server.unhandled_requests`](#metric-httpserverunhandled_requests)
+
+<!-- tocstop -->
+
+## HTTP client
+
+All Http client metrics are reported by `System.Net.Http` meter.
+
+### Metric: `http.client.connections.usage`
+
+**Status**: [Experimental][DocumentStatus]
+
+This metric is required.
+
+<!-- semconv metric.dotnet.http.client.connections.usage(metric_table) -->
+| Name     | Instrument Type | Unit (UCUM) | Description    |
+| -------- | --------------- | ----------- | -------------- |
+| `http.client.connections.usage` | UpDownCounter | `{connection}` | Number of outbound HTTP connections that are currently active or idle on the client [1] |
```

Comment:
I think the usage aspect will not be that useful in practice. Its going to generate a lot of noise. 

With the connection duration counter below, I'd suggest removing this counter altogether.

Reply:
hm, it seems the usage here represents a number of open connections  - the `connection.duration` histogram does not provide the number of active/idle connections, it's only reported after the connection is closed, we can derive connection creation and close rate from it.

I.e. if we don't introduce it now:
- we won't have a way to find how many connections are open at each moment
- it seems common and useful to measure the number of open connections
- we can still add it in .NET 9 if we realize it's useful

we can rename it to `http.client.open_connections`.

I believe knowing the number of idle connections from metrics is important - alerting on logs is a questionable and expensive thing to do. Retrieving logs when dealing with incidents in production is slow and unreliable - logs can be files on a file system.

Reply:
BTW I think usage is used incorrectly here - usage assumes there is a limit value reported and the sum of all state dimensions is equal to the limit. This is not the case here.

Reply:
Updated this PR with `http.client.open_connections` - PTAL. We should so consider prefixing `connection.state` with something - I assume different connections can have different set of states and we should not mix them - e.g. there is no idle or active states here - https://learn.microsoft.com/en-us/dotnet/api/system.data.connectionstate?view=net-7.0&viewFallbackFrom=dotnet-plat-ext-7.0

Reply:
I'd be happy not to use it :) Any of these seems fine to me:
http.client.open_connections
http.client.connections.count
http.client.connections

If I was picking my preference I'd probably go with the last one but its splitting hairs for me. I think its very reasonable to still have a state attribute that marks connections as idle or not.

@samsp-msft and @antonfirsov can of course say otherwise, but for me (and I am guessing for them), the aesthetic dislike really came from "usage" not having a clear and intuitive meaning. If I told someone "the value of the connection.usage metric is 26" they will probably give me weird confused looks. If I tell them "the value of the open_connections metric is 26" they probably have an intuitive understanding what that means.

Reply:
we had a chat with otel semconv WG members today and thinking about the following choice:

1. `*.connection_pool.usage` + `*.conneciton_pool.limit` - which tells e.g. that there are 5 active and 25 idle connections in the connection pool of size 30

OR

2. `*.connections.count` if there is no pool or limit is not known

I guess the dislike from `usage` came from not having a limit reported and not talking about the connection pool. To connection pool usage seems intuitive.

Let me know what your thoughts are and what would be a preferred option here